src/app/state/orderInputSlice.ts

```diff
@@ -57,114 +60,160 @@ export interface TokenInput {
+    LIMIT: postOnly ? adex.OrderType.POSTONLY : adex.OrderType.LIMIT,
+  }[dexterOrderType];
+}
+
+export function toDexterOrderType(adexOrderType: adex.OrderType): OrderType {
+  if (adexOrderType === adex.OrderType.MARKETPARTIAL) {
+    throw new Error("Unimplemented");
   }
-);
+  return {
+    MARKET: OrderType.MARKET,
+    LIMIT: OrderType.LIMIT,
+    POSTONLY: OrderType.LIMIT,
+  }[adexOrderType];
+}
+
+function swapSideIfToken2Specified(
+  side: OrderSide,
+  specifiedToken: SpecifiedToken
+): OrderSide {
+  if (specifiedToken === SpecifiedToken.UNSPECIFIED) {
+    throw new Error("Token must be specified");
+  }
+  const swapSide = (side: OrderSide) => {
+    return side === OrderSide.BUY ? OrderSide.SELL : OrderSide.BUY;
+  };
+  return specifiedToken === SpecifiedToken.TOKEN_2 ? swapSide(side) : side;
+}
+
+// Map the user-facing side to the Alphadex convention. Alphadex defines
+// "side" relative to the specifiedToken, while DeXter sets it based on
+// token order of the pairAddress. The general rule is, whenever TOKEN_2
+// is specified, the side needs to be swapped.
+// |------------|----------------|----------|
+// | DexterSide | specifiedToken | AdexSide |
+// |------------|----------------|----------|
+// | buy        | token1         | buy      |
+// | buy        | token2         | sell     | <- swap
+// | sell       | token1         | sell     |
+// | sell       | token2         | buy      | <- swap
+// |------------|----------------|----------|
+export function toAdexOrderSide(
+  dexterSide: OrderSide,
+  specifiedToken: SpecifiedToken
+): adex.OrderSide {
+  return swapSideIfToken2Specified(dexterSide, specifiedToken);
+}
+
+// Convert Alphadex side to user-facing side in DeXter. DeXter defines
+// "side" based on token order in pairAddress, while Alphadex sets it based
+// on the specifiedToken.
+export function toDexterOrderSide(
+  adexSide: adex.OrderSide,
+  specifiedToken: SpecifiedToken
+): OrderSide {
+  // same logic to convert back, so we can reuse the same logic
+  return swapSideIfToken2Specified(adexSide, specifiedToken);
+}
+
+export function noValidationErrors(
+  validationPrice: ValidationResult,
+  validationToken1: ValidationResult,
+  validationToken2: ValidationResult
+): boolean {
+  return (
+    validationPrice.valid && validationToken1.valid && validationToken2.valid
+  );
+}
+
+export function pairAddressIsSet(pairAddress: string): boolean {
+  return pairAddress !== "";
+}
+
+export function priceIsValid(price: number, type: OrderType): boolean {
+  const isMarketOrder = type === OrderType.MARKET;
+  const isValidLimitOrder = type === OrderType.LIMIT && price > 0;
+  return isMarketOrder || isValidLimitOrder;
+}
+
+export function tokenIsSpecified(specifiedToken: SpecifiedToken): boolean {
+  return specifiedToken !== SpecifiedToken.UNSPECIFIED;
+}
 
 // for getting balances out of pairSelector slice
+// TODO(dcts): ask @chaotic whether this can live inside pairSelector now that
```

Comment:
yes, sure, and probably it will live in the `appSlice` after Fred's refactoring of the state